README.md
# WebApiClient

**WebApiClient** is an application framework aimed at making RESTful web service
requests in a standardized way. It works by configuring named routes for each web
service endpoint so your application focuses on consuming the logical web service
API, not HTTP implementation details.

The project is divided into modules, staring with a **Core** module and then
branching off into various supporting modules that add functionality like [object
mapping](#module-restkit), [caching](#module-cache), and [UI support](#module-ui).

# Module: Core

The **Core** module provides a HTTP client framework based on _routes_ with support
for _object mapping_ for transforming requests and responses between native objects
and serialized forms, such as JSON. This module provides just a protocol based API
and some scaffolding classes to support the API, but does not provide an actual full
implementation itself, so that different HTTP back-ends can be used as needed. The
[AFNetworking module](#module-afnetworking) provides a full implementation of the API.

The
[WebApiClient][WebApiClient] protocol defines the main HTTP client entry point for
applications to use. The API is purposefully simple and based on asynchronous block
callbacks:

```objc
- (void)requestAPI:(NSString *)name
 withPathVariables:(id)pathVariables
        parameters:(id)parameters
              data:(id<WebApiResource>)data
		  finished:(void (^)(id<WebApiResponse> response, NSError *error))callback;
```

An example invocation of this API might look like this:

```objc
// make a GET request to /documents/123
[client requestAPI:@"doc" withPathVariables:@{@"uniqueId" : @123 } parameters:nil data:nil
          finished:^(id<WebApiResponse> response, NSError *error) {
	if ( !error ) {
		MyDocument *doc = response.responseObject;
	} else if ( response.statusCode == 422 ) {
		// handle 422 (validation) errors here...
	}
}];
```

## Background callback support

By default the callback block is called on the main thread (queue). If you prefer to
have the callabck on a specific queue, you can use an alternate method that accepts a
dispatch queue as a parameter. In that case, the passed in queue will be used for the
callback:

```objc
- (void)requestAPI:(NSString *)name
 withPathVariables:(id)pathVariables
        parameters:(id)parameters
              data:(id<WebApiResource>)data
             queue:(dispatch_queue_t)callbackQueue
          progress:(nullable WebApiClientRequestProgressBlock)progressCallback
		  finished:(void (^)(id<WebApiResponse> response, NSError *error))callback;
```

## Progress callback support

The same method that accepts an explicit callback block shown in the previous section
also accepts an optional `WebApiClientRequestProgressBlock`, which is defined as this:

```objc
typedef void (^WebApiClientRequestProgressBlock)(NSString *routeName,
                                                 NSProgress * _Nullable uploadProgress,
                                                 NSProgress * _Nullable downloadProgress);
```

By passing in this type of block to the `progress` parameter, you can monitor both
the upload and download progress of the HTTP request. In addition the
`WebApiClientRequestDidProgressNotification` and
`WebApiClientResponseDidProgressNotification` notifications can be used to listen for
progress updates as well. The `WebApiClientProgressNotificationKey` notification user
info key will contain the relevant `NSProgress` object.

## Synchronous request support

Sometimes it can be useful to make a blocking, synchronous request to get a HTTP
resource. WebApiClient supports that as well:

```objc
- (id<WebApiResponse>)blockingRequestAPI:(NSString *)name
                       withPathVariables:(id)pathVariables
                              parameters:(id)parameters
                                    data:(id<WebApiResource>)data
                             maximumWait:(NSTimeInterval)maximumWait
                                   error:(NSError **)error;
```

Calling this method will block the calling thread until either the response is
available or `maximumWait` seconds have elapsed.

## Routing

The [WebApiRoute][WebApiRoute] protocol defines a single API endpoint definition,
assigned a unique name. Routes are typically configured when an application starts
up. Each route defines some standardized properties, such as a HTTP `method` and URL
`path`. For convenience, routes support arbitrary property access via Objective-C's
keyed subscript support, so the following is possible:

```objc
id<WebApiRoute> myRoute = ...;

// access the path property
NSString *path1 = myRoute.path;

// access the path property using keyed subscript notation
NSString *path2 = myRoute[@"path"];

// access some arbitrary property not defined in WebApiRoute specifically
id something = myRoute[@"extendedProperty"];
```

For even more convenience, WebApiRoute provides
[extensions][NSDictionary-WebApiRoute] to `NSDictionary` and `NSMutableDictionary` so
that they conform to `WebApiRoute` and `MutableWebApiRoute`, respectively. That means
you can use dictionaries directly as routes, like this:

```objc
// define a route
id<WebApiRoute> myRoute = @{ @"name" : "login", @"path" : @"user/login", @"method" : @"POST" };

// create a mutable copy and extend
id<MutableWebApiRoute> mutableRoute = [myRoute mutableCopy];
mutableRoute[@"extendedProperty"] = @"special";
```

## Object mapping

The [WebApiDataMapper][WebApiDataMapper] protocol defines an API for _encoding_
native objects into HTTP requests and _mapping_ HTTP responses into native objects.
Routes can be configured with a `dataMapper` property to support this feature. The
API is also pretty simple:

```objc
@protocol WebApiDataMapper <NSObject>

// Map a source data object into some domain object.
- (id)performMappingWithSourceObject:(id)sourceObject route:(id<WebApiRoute>)route error:(NSError *__autoreleasing *)error;

// Encode a domain object into an encoded form, such as @c NSDictionary or @c NSData.
- (id)performEncodingWithObject:(id)domainObject route:(id<WebApiRoute>)route error:(NSError *__autoreleasing *)error;

@end
```

# Module: AFNetworking

The **AFNetworking** module provides a full implementation of the `WebApiClient` API,
based on [AFNetworking][afn] and `NSURLSession`.

## Route configuration

Routes can be configured in code via the `registerRoute:forName:` method, but more
conveniently they can be configured via [BREnvironment][brenv]. The `webservice.api`
key will be inspected by default, and can be a dictionary representing all the routes
that should be registered for the application. For example, the following JSON would
register three routes, `login`, `register`, and `absolute`:

```json
{
  "App_webservice_protocol" : "https",
  "App_webservice_host" : "example.com",
  "App_webservice_port" : 443,
  "webservice" : {
    "api" : {
      "register" : {
        "method" : "POST",
        "path" : "user/register",
      },
      "login" : {
        "method" : "POST",
        "path" : "user/login",
      },
      "absolute" : {
        "method" : "GET",
        "path" : "https://example.com/something"
      }
    }
  }
}
```

You'll notice that the `register` and `login` routes have relative paths. All
webservice URLs are constructed as relative to a configurable `baseApiURL` property,
which by default is configured via the various `App_webservice_*`
[BREnvironment][brenv] keys you can see in the previous example JSON.

### GZip compression support

The `gzip` property on routes is supported. When set to `true` any _request_ data
will be compressed and a request HTTP header of `Content-Encoding: gzip` will be
added.

To support compressed _response_ data (highly recommended!) you only need to
configure an `Accept-Encoding: gzip` HTTP header, either individually on routes via
the `requestHeaders` property or via the `globalHTTPRequestHeaders` property
available on `AFNetworkingWebApiClient`.

Here's an example route that configures both request and response compression:

```json
{
  "webservice" : {
    "api" : {
      "trim" : {
        "method" : "POST",
        "path" : "upload/jumbo",
        "gzip" : true,
        "requestHeaders" : {
          "Accept-Encoding" : "gzip"
        }
      }
    }
  }
}
```

### Upload raw data

Raw data can be uploaded directly in the body of the HTTP request. This can be
useful, for example, when you need to upload images, or any other type of data, and
the URL contains sufficient information to identify the content. To perform a raw
data upload, pass a [WebApiResource][WebApiResource] instance on the `data` parameter
in the WebApiClient API. WebApiClient provides two implementations of
`WebApiResource`: `DataWebApiResource` for in-memory data and `FileWebApiResource`
for file-based data. The `WebApiResource` instance you pass in will be sent directly
in the body of the request, and appropriate `Content-Type` and `Content-MD5` HTTP
headers will be included. This means the `parameters` object is ignored. If you need
to post both parameters _and_ a file, use the `multipart/form-data` upload method
described in the next section.

### Upload data (multipart/form-data)

Instead of uploading raw data, you can also upload using a `multipart/form-data`
attachment encoding by passing a [WebApiResource][WebApiResource] instance on the
`data` parameter in the WebApiClient API and configuring the route with a
serialization type of `form`, like this:

```json
{
  "webservice" : {
    "api" : {
      "trim" : {
        "method" : "POST",
        "path" : "upload/image",
        "serializationName" : "form"
      }
    }
  }
}
```

The WebApiClient API `parameters` object, if provided, will also be included in the
request, serialized into additional parts of the request body.

### Download raw data

You can configure a route to save the response data into a file, instead of the
default of loading the response in RAM, by adding a `saveAsResource` property with a
_truthy_ value, like this:

```json
{
  "webservice" : {
    "api" : {
      "download" : {
        "method" : "GET",
        "path" : "download/image",
        "saveAsResource" : true
      }
    }
  }
}
```

Then the `responseObject` returned in the `WebApiResponse` will be a
[WebApiResource][WebApiResource] which you can then move to an appropriate location
as needed, for example:

```objc
[client requestAPI:@"download" withPathVariables:nil parameters:nil data:nil
             queue:dispatch_get_main_queue()
          progress:nil
          finished:^(id<WebApiResponse> response, NSError *error) {
    if ( !error ) {
        id<WebApiResource> resource = response.responseObject;
        NSURL *dest = [NSURL fileURLWithPath:@"/some/path"];
        [[NSFileManager defaultManager] moveItemAtURL:[resource URLValue] toURL:dest error:nil];
    }
}];
```


# Module: Cache

The **Cache** module provides response caching support to the `WebApiClient` API by
providing the [`PINCacheWebApiClient`][PINCacheWebApiClient] proxy  that can cache
result objects using [PINCache][pcache]. Caching support is enabled by configuring a
`cacheTTL` property on routes, for example:

```json
{
  "webservice" : {
    "api" : {
      "info" : {
        "method" : "GET",
        "path" : "infrequentlyupdated/info",
        "cacheTTL" : 3600
      }
    }
  }
}
```

See the [CachingWebApiRoute][CachingWebApiRoute] protocol for more route cache
details.

## Routes that invalidate cached data for other routes

You can also configure a route so that it invalidates any cached data for _other_
routes. A good example of where this is useful is when you define a _list_ route that
returns a list of objects, and another _add_ route to add to that same list of
objects. We can make the latter route invalidate the cached data of the former like
this:

```json
{
  "webservice" : {
    "api" : {
      "list" : {
        "method" : "GET",
        "path" : "stuff/list",
        "cacheTTL" : 3600
      },
      "add" : {
        "method" : "PUT",
        "path" : "stuff/:thingId",
        "invalidatesCachedRouteNames" : [ "list" ]
      }
    }
  }
}
```

The `invalidatesCachedRouteNames` is configured as an array of route names that
should be invalidated when that route is called successfully.

## Ignoring route URL query parameters

By default URL query parameters will be included when calculating the cache key for
each route. Sometimes it can be useful to ignore the query parameters, however. For
example, pre-signed Amazon S3 resource URLs contain authorization query parameters
that change each time the same resource is requested. By configuring the route with
`cacheIgnoreQueryParameters = YES` then the query parameters for that route will be
**not** be included in the request's cache key:

```json
{
  "webservice" : {
    "api" : {
      "info" : {
        "method" : "GET",
        "path" : "https://s3.amazon.com/info/foo.txt",
        "cacheTTL" : 3600,
        "cacheIgnoreQueryParameters" : true
      }
    }
  }
}
```

## Checking for cached data

The [`CachingWebApiClient`][CachingWebApiClient] API adds a method that can be used
for testing if cached data is available:

```objc
- (void)requestCachedAPI:(NSString *)name
       withPathVariables:(nullable id)pathVariables
              parameters:(nullable id)parameters
                   queue:(dispatch_queue_t)callbackQueue
                finished:(void (^)(id<WebApiResponse> _Nullable response, NSError * _Nullable error))callback;
```

The callback will be passed a response only if the data was available in the cache.
Sometimes its handy to know if something is already downloaded!


# Module: RestKit

The **RestKit** module provides an _object mapping_ implementation for the
`WebApiClient` API based on the [RestKit][rk]. It provides a way to transform native
objects into JSON, and vice versa. This module only makes use of the
`RestKit/ObjectMapping` module, so it does not conflict with AFNetworking 2. In fact,
part of the motivation for WebApiClient was to be able to use AFNetworking 2 with
RestKit's object mapping support because RestKit's networking layer is based on
AFNetworking 1. In some respects the WebApiClient API provides some of the same
scaffolding that the full RestKit project provides.

## Mapping configuration

The [`RestKitWebApiDataMapper`][RestKitWebApiDataMapper] class supports a shared
singleton pattern that your application can configure when it starts up with any
required `RKObjectMapping` objects. You configure it like this:

```objc
RestKitWebApiDataMapper *dataMapper = [RestKitWebApiDataMapper sharedDataMapper];

// get RestKit mapper for user objects
RKObjectMapper *userObjectMapper = ...;

// register user mapper for requests and responses
[dataMapper registerRequestObjectMapping:[userObjectMapper inverseMapping] forRouteName:@"login"];
[dataMapper registerResponseObjectMapping:userObjectMapper forRouteName:@"login"];
```

## Block-based encoding & mapping

The [`RestKitWebApiDataMapper`][RestKitWebApiDataMapper] class also supports
block-based mapping hooks for both request encoding and response mapping. The blocks
are executed _after_ any configured `RKObjectMapper` has done its job on the request
or response data.

One useful example of this support is for wiring up parent-child relationship
properties that are implied by the data. Imagine a `Person` class that has an array
of child `Person` objects, and each child has a `parent` property that points to its
parent `Person` instance:

```objc
@interface Person : NSObject

@property (strong) NSString *name;
@property (strong) NSArray<Person *> *children;
@property (weak) Person *parent;

@end
```

The server returns JSON like this:

```json
{
  "name" : "John Doe",
  "children" : [
    { "name" : "Johnny Doe" },
    { "name" : "Jane Doe" }
  ]
}
```

To populate each child's `parent` property with the **John Doe** object, a response
mapping block could be configured like this:

```objc
[dataMapper registerResponseMappingBlock:^(id sourceObject, id<WebApiRoute> route, NSError * __autoreleasing *error) {
	if ( [sourceObject isKindOfClass:[Person class]] ) {
		// populate the Child -> Parent relationship
		Person *parent = sourceObject;
		for ( Person *child in parent.children ) {
			child.parent = parent;
		}
	}
	return sourceObject;
} forRouteName:@"parent-child-tree"];
```


## Route configuration

To use RestKit-based object mapping with a route, you configure the `dataMapper`
property of the route with `RestKitWebApiDataMapper` like this:

```json
{
  "webservice" : {
    "api" : {
      "login" : {
        "method" : "POST",
        "path" : "user/login",
        "dataMapper" : "RestKitWebApiDataMapper"
      }
    }
  }
}
```

Sometimes the request or response JSON needs to be nested in some top-level object.
For example imagine that the register endpoint expects the user object to be posted
as JSON like this:

```json
{
  "user" : { "email" : "joe@example.com", "name" : "Joe" }
}
```

This can be done by adding a `dataMapperRequestRootKeyPath` property (or
`dataMapperResponseRootKeyPath` for mapping responses), like this:

```json
{
  "webservice" : {
    "api" : {
      "login" : {
        "method" : "POST",
        "path" : "user/login",
        "dataMapper" : "RestKitWebApiDataMapper",
        "dataMapperRequestRootKeyPath" : "user"
      }
    }
  }
}
```


# Module: UI

The **UI** module provides some UI utilities, such as the
[`WebApiClientActivitySupport`][WebApiClientActivitySupport] class that listens to
route requests and for those that specify `preventUserInteraction` with a truthy
value will throw up a full-screen "request taking too long" view to let the user of
the app know it's waiting for a response. For example, a route can be configured like:

```json
{
  "webservice" : {
    "api" : {
      "login" : {
        "method" : "POST",
        "path" : "user/login",
        "preventUserInteraction" : true
      }
    }
  }
}
```


 [cocoapods]: https://cocoapods.org/
 [brenv]: https://github.com/Blue-Rocket/BREnvironment
 [afn]: https://github.com/AFNetworking/AFNetworking
 [rk]: https://github.com/RestKit/RestKit
 [pcache]: https://github.com/pinterest/PINCache
 [CachingWebApiClient]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient-Cache/CachingWebApiClient.h
 [CachingWebApiRoute]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient-Cache/CachingWebApiRoute.h
 [NSDictionary-WebApiRoute]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient/NSDictionary%2BWebApiClient.h
 [PINCacheWebApiClient]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient-Cache/PINCacheWebApiClient.h
 [RestKitWebApiDataMapper]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient-RestKit/RestKitWebApiDataMapper.h
 [WebApiClient]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient/WebApiClient.h
 [WebApiClientActivitySupport]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient-UI/WebApiClientActivitySupport.h
 [WebApiDataMapper]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient/WebApiDataMapper.h
 [WebApiResource]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient/WebApiResource.h
 [WebApiRoute]: https://github.com/Blue-Rocket/WebApiClient/blob/master/WebApiClient/Code/WebApiClient/WebApiRoute.h


CHANGELOG.md
# Change Log

## [1.2.1](https://github.com/Blue-Rocket/WebApiClient/tree/1.2.1) (2016-09-15)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.2.0...1.2.1)

**Closed issues:**

- NSInternalInconsistencyException thrown if response does not contain configured root key path [\#5](https://github.com/Blue-Rocket/WebApiClient/issues/5)

## [1.2.0](https://github.com/Blue-Rocket/WebApiClient/tree/1.2.0) (2016-09-07)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.1.3...1.2.0)

**Implemented enhancements:**

- RestKit 0.27 fails to compile  [\#4](https://github.com/Blue-Rocket/WebApiClient/issues/4)

## [1.1.3](https://github.com/Blue-Rocket/WebApiClient/tree/1.1.3) (2016-09-07)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.1.2...1.1.3)

## [1.1.2](https://github.com/Blue-Rocket/WebApiClient/tree/1.1.2) (2016-06-20)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.1.1...1.1.2)

## [1.1.1](https://github.com/Blue-Rocket/WebApiClient/tree/1.1.1) (2016-06-20)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.1.0...1.1.1)

**Implemented enhancements:**

- Support encoding ObjC nil values into JSON null [\#3](https://github.com/Blue-Rocket/WebApiClient/issues/3)

## [1.1.0](https://github.com/Blue-Rocket/WebApiClient/tree/1.1.0) (2016-06-20)
[Full Changelog](https://github.com/Blue-Rocket/WebApiClient/compare/1.0.0...1.1.0)

**Implemented enhancements:**

- Add support for multi-user cache [\#2](https://github.com/Blue-Rocket/WebApiClient/issues/2)

**Fixed bugs:**

- Content-MD5 header value should be Base64 encoded [\#1](https://github.com/Blue-Rocket/WebApiClient/issues/1)



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

WebApiClient/Code/WebApiClient/WebApiClientSupport.m
//
//  WebApiClientSupport.m
//  WebApiClient
//
//  Created by Matt on 12/08/15.
//  Copyright (c) 2015 Blue Rocket. Distributable under the terms of the Apache License, Version 2.0.
//

#import "WebApiClientSupport.h"

#import <BRCocoaLumberjack/BRCocoaLumberjack.h>
#import <BREnvironment/BREnvironment.h>
#import <BRLocalize/Core.h>
#import <MAObjCRuntime/MARTNSObject.h>
#import <MAObjCRuntime/RTProperty.h>
#import <SOCKit/SOCKit.h>
#import "WebApiAuthorizationProvider.h"
#import "WebApiClientEnvironment.h"
#import "WebApiDataMapper.h"

NSString * const WebApiClientSupportAppApiKeyDefaultHTTPHeaderName = @"X-App-API-Key";
NSString * const WebApiClientSupportAppIdDefaultHTTPHeaderName = @"X-App-ID";

static NSString * const kRoutePropertyPattern = @"_pattern";
static NSString * const kRoutePropertyDataMapperInstance = @"_dataMapper";

@implementation WebApiClientSupport {
	NSMutableDictionary *routes;
	NSURL *baseApiURL;
}

- (id)init {
	return [self initWithEnvironment:[BREnvironment sharedEnvironment]];
}

- (id)initWithEnvironment:(BREnvironment *)theEnvironment {
	if ( (self = [super init]) ) {
		routes = [[NSMutableDictionary alloc] initWithCapacity:16];
		self.appApiKey = theEnvironment[WebApiClientSupportAppApiKeyEnvironmentKey];
		self.appApiKeyHTTPHeaderName = WebApiClientSupportAppApiKeyDefaultHTTPHeaderName;
		self.appIdHTTPHeaderName = WebApiClientSupportAppIdDefaultHTTPHeaderName;
		self.appId = [[NSBundle mainBundle] objectForInfoDictionaryKey:(id)kCFBundleIdentifierKey];
		[self loadDefaultRoutes:theEnvironment];
		baseApiURL = [self setupBaseApiURL:theEnvironment];
	}
	return self;
}

+ (instancetype)sharedClient {
	static WebApiClientSupport *shared;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		shared = [[self alloc] init];
	});
	return shared;
}

- (void)loadDefaultRoutes:(BREnvironment *)environment {
	// look for routes in config.json/webservice.api dictionary
	id routeConfigs = environment[@"webservice.api"];
	if ( ![routeConfigs conformsToProtocol:@protocol(NSFastEnumeration)] ) {
		return;
	}
   for ( id routeName in routeConfigs ) {
	   DDLogDebug(@"Inspecting web api route %@", routeName);
	   [self registerRoute:routeConfigs[routeName] forName:routeName];
   }
}

- (NSURL *)setupBaseApiURL:(BREnvironment *)environment {
	NSString *protocol = environment[WebApiClientSupportServerProtocolEnvironmentKey];
	NSString *host = environment[WebApiClientSupportServerHostEnvironmentKey];
	int port = [environment[WebApiClientSupportServerPortEnvironmentKey] intValue];
	
	if ( !port
		|| ([protocol isEqualToString:@"http"] && port == 80 )
		|| ([protocol isEqualToString:@"https"] && port == 443) ) {
		// don't include port in base URL
		return [NSURL URLWithString:[NSString stringWithFormat:@"%@://%@/", protocol, host]];
	}
	return [NSURL URLWithString:[NSString stringWithFormat:@"%@://%@:%d/", protocol, host, port]];
}

- (void)registerRoute:(id<WebApiRoute>)route forName:(NSString *)name {
	NSMutableDictionary *internalRoute = [[NSMutableDictionary alloc] initWithCapacity:4];
	if ( [route isKindOfClass:[NSDictionary class]] ) {
		// copy all route props
		[internalRoute addEntriesFromDictionary:(NSDictionary *)route];
	} else {
		// manually copy WebApiRoute values
		internalRoute[NSStringFromSelector(@selector(method))] = route.method;
		internalRoute[NSStringFromSelector(@selector(path))] = route.path;
	}
	
	// force name to given name
	internalRoute.name = name;
	
	// add our SOCPattern
	internalRoute[kRoutePropertyPattern] = [[SOCPattern alloc] initWithString:route.path];
	
	routes[name] = internalRoute;
	
	// extending classes may want to do more here, just please call [super registerRoute:forName:]
}

- (id<WebApiRoute>)routeForName:(NSString *)name error:(NSError *__autoreleasing *)error {
	id<WebApiRoute> result = routes[name];
	if ( !result && error ) {
		*error = [NSError errorWithDomain:WebApiClientErrorDomain code:WebApiClientErrorRouteNotAvailable userInfo:
				  @{@"name" : name, NSLocalizedDescriptionKey : [@"{web.api.missingRoute}" localizedString] }];
	}
	return result;
}

- (NSURL *)baseApiURL {
	return baseApiURL;
}

- (NSDictionary *)dictionaryForParametersObject:(id)parameters {
	if ( [parameters isKindOfClass:[NSDictionary class]] ) {
		return parameters;
	}
	
	static NSSet *ignoreProperties;
	if ( !ignoreProperties ) {
		ignoreProperties = [NSSet setWithArray:@[@"hash", @"superclass", @"description", @"debugDescription"]];
	}
	
	// inspect properties of object up to, but not including, root class, and return dictionary of all those values
	NSMutableArray *propNames = [[NSMutableArray alloc] initWithCapacity:8];
	Class objClass = [parameters class];
	while ( class_getSuperclass(objClass) ) {
		for ( RTProperty *prop in [objClass rt_properties] ) {
			if ( [ignoreProperties containsObject:prop.name] ) {
				continue;
			}
			[propNames addObject:prop.name];
		}
		objClass = class_getSuperclass(objClass);
	}
	NSDictionary *result = [parameters dictionaryWithValuesForKeys:propNames];
	NSSet *nulls = [result keysOfEntriesPassingTest:^BOOL(id key, id obj, BOOL *stop) {
		return [obj isKindOfClass:[NSNull class]];
	}];
	if ( [nulls anyObject] ) {
		NSMutableDictionary *mutableResult = [result mutableCopy];
		[mutableResult removeObjectsForKeys:[nulls allObjects]];
		result = mutableResult;
	}
	return result;
}

- (NSURL *)URLForRoute:(id<WebApiRoute>)route pathVariables:(id)pathVariables parameters:(id)parameters error:(NSError *__autoreleasing *)error {
	NSString *path = nil;
	SOCPattern *pattern = route[kRoutePropertyPattern];
	if ( pattern ) {
		path = [pattern stringFromObject:pathVariables withBlock:^NSString *(NSString *propertyValue) {
			return [propertyValue stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
		}];
	} else {
		path = route.path;
	}
	
	if ( !path ) {
		DDLogError(@"No API path defined for route %@", route);
		if ( error ) {
			*error = [NSError errorWithDomain:WebApiClientErrorDomain code:WebApiClientErrorRouteNotAvailable userInfo:
					  @{@"route" : route, NSLocalizedDescriptionKey : [@"{web.api.missingRoutePath}" localizedString] }];
		}
		return nil;
	}
	
	if ( parameters ) {
		NSString *params = [[self dictionaryForParametersObject:parameters] asURLQueryParameterString];
		if ( [params length] > 0 ) {
			path = [path stringByAppendingFormat:@"?%@", params];
		}
	}
	
	NSURL *url = [NSURL URLWithString:path relativeToURL:[self baseApiURL]];
	return url;
}

- (id<WebApiDataMapper>)dataMapperForRoute:(id<WebApiRoute>)route {
	id<WebApiDataMapper> mapper = route[kRoutePropertyDataMapperInstance];
	if ( mapper ) {
		return mapper;
	}
	NSString *dataMapperName = route.dataMapper;
	if ( [dataMapperName length] < 1 ) {
		return nil;
	}
	// is this a class name?
	Class dataMapperClass = NSClassFromString(dataMapperName);
	if ( dataMapperClass ) {
		if ( [dataMapperClass conformsToProtocol:@protocol(WebApiSingletonDataMapper)] ) {
			mapper = [dataMapperClass sharedDataMapper];
		} else {
			mapper = [[dataMapperClass alloc] init];
		}
	}
	if ( mapper && [route conformsToProtocol:@protocol(MutableWebApiRoute)] ) {
		route[kRoutePropertyDataMapperInstance] = mapper;
	}
	return mapper;
}

- (void)requestAPI:(NSString *)name withPathVariables:(id)pathVariables parameters:(NSDictionary *)parameters data:(id<WebApiResource>)data
		  finished:(void (^)(id<WebApiResponse>, NSError *))callback {
	[self requestAPI:name withPathVariables:pathVariables parameters:parameters data:data queue:dispatch_get_main_queue() progress:nil finished:callback];
}

- (void)requestAPI:(NSString *)name withPathVariables:(id)pathVariables parameters:(id)parameters data:(id<WebApiResource>)data
			 queue:(dispatch_queue_t)callbackQueue
		  progress:(nullable WebApiClientRequestProgressBlock)progressCallback
		  finished:(nonnull void (^)(id<WebApiResponse> _Nonnull, NSError * _Nullable))callback {
	// extending classes probably want to do something useful here
}

- (nullable id<WebApiResponse>)blockingRequestAPI:(NSString *)name
								withPathVariables:(nullable id)pathVariables
									   parameters:(nullable id)parameters
											 data:(nullable id<WebApiResource>)data
									  maximumWait:(const NSTimeInterval)maximumWait
											error:(NSError **)error {
	
	// results
	__block id<WebApiResponse> clientResponse = nil;
	__block NSError *clientError = nil;
	
	// we're going to block the calling thread here, for up to maximumWait seconds
	NSCondition *condition = [NSCondition new];
	[condition lock];
	
	dispatch_queue_t bgQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
	__block BOOL finished = NO;
	[self requestAPI:name withPathVariables:pathVariables parameters:parameters data:data queue:bgQueue progress:nil finished:^(id<WebApiResponse>  _Nonnull response, NSError * _Nullable error) {
		[condition lock];
		finished = YES;
		clientResponse = response;
		clientError = error;
		[condition signal];
		[condition unlock];
	}];
	
	// block and wait for our response now...
	BOOL timeout = NO;
	if ( maximumWait > 0 ) {
		NSDate *timeoutDate = [NSDate dateWithTimeIntervalSinceNow:maximumWait];
		while ( !finished && [timeoutDate timeIntervalSinceNow] > 0 ) {
			timeout = ![condition waitUntilDate:timeoutDate];
		}
	} else {
		while ( !finished ) {
			[condition wait];
		}
	}
	[condition unlock];
	
	if ( timeout && !clientError ) {
		log4Warn(@"No response returned from route %@ within %0.1f seconds", name, maximumWait);
		NSString *message = [NSString stringWithFormat:[@"{web.api.responseTimeout}" localizedString], @(maximumWait)];
		clientError = [NSError errorWithDomain:WebApiClientErrorDomain code:WebApiClientErrorResponseTimeout userInfo:
					   @{@"name" : name, NSLocalizedDescriptionKey : message }];
	}
	
	if ( error && clientError ) {
		*error = clientError;
	}
	return clientResponse;
}

- (void)addRequestHeadersToRequest:(NSMutableURLRequest *)request forRoute:(id<WebApiRoute>)route {
	NSDictionary<NSString *, NSString *> *routeHeaders = route.requestHeaders;
	[self.globalHTTPRequestHeaders enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
		if ( !routeHeaders[key] ) {
			[request addValue:obj forHTTPHeaderField:key];
		}
	}];
	[routeHeaders enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
		[request addValue:obj forHTTPHeaderField:key];
	}];
}

- (void)addAuthorizationHeadersToRequest:(NSMutableURLRequest *)request forRoute:(id<WebApiRoute>)route {
	if ( self.appApiKey ) {
		[request setValue:self.appApiKey forHTTPHeaderField:self.appApiKeyHTTPHeaderName];
	}
	if ( self.appId ) {
		[request setValue:self.appId forHTTPHeaderField:self.appIdHTTPHeaderName];
	}
	if ( self.authorizationProvider ) {
		[self.authorizationProvider configureAuthorizationForRoute:route request:request];
	}
}

- (NSArray<NSHTTPCookie *> *)cookiesForAPI:(nullable NSString *)name inCookieStorage:(nullable NSHTTPCookieStorage *)cookieJar {
	NSMutableArray *result = [[NSMutableArray alloc] initWithCapacity:8];
	NSURL *cookieURL = ([name length] ? [self URLForRoute:[self routeForName:name error:nil] pathVariables:nil parameters:nil error:nil] : baseApiURL);
	NSHTTPCookieStorage *jar = (cookieJar ? cookieJar : [NSHTTPCookieStorage sharedHTTPCookieStorage]);
	if ( cookieURL ) {
		NSArray *found = [jar cookiesForURL:baseApiURL];
		if ( found ) {
			[result addObjectsFromArray:found];
		}
	}
	return result;
}

@end
